sdk/client.py

```diff
@@ -147,20 +43,56 @@ def __init__(
         token_provider: Optional[TokenProvider] = None,
         endpoints: Optional[Endpoints] = None,
     ):
-        if not (username and password) and not token_provider:
-            raise Exception(
-                "At least a username/password combination or"
-                "TokenProvider object should be provided."
+        if not username and not token_provider:
+            raise ValueError(
+                "At least a username or TokenProvider object should be provided."
             )
-        self.endpoints = endpoints or Endpoints()
-        self.group_id = group_id
+        self._check_token_provider(token_provider)
 
-        if username and password:
-            token_provider = UsernamePasswordTokenProvider(
-                username, password, f"{self.endpoints.account}/api/v1/auth/login"
-            )
+        if username:
+            token_provider = self._credential_login(username, password)
 
         self.authenticator = HTTPBearerAuthenticator(token_provider)
+        self.endpoints = self._make_endpoints(endpoints)
+        self.group_id = group_id
+
+    @staticmethod
+    def _make_endpoints(endpoints: Optional[Endpoints]) -> Endpoints:
+        if endpoints is None:
+            return Endpoints()
+
+        if not isinstance(endpoints, Endpoints):
+            raise ValueError("Endpoints must be a Endpoints instance")
+
+        return endpoints
+
+    @staticmethod
+    def _check_token_provider(token_provider: Optional[TokenProvider]) -> None:
+        if not token_provider:
+            return
+
+        err = ValueError("token_provider must be a TokenProvider subclass")
+        try:
+            # The type ignore is because I wouldn't know how to fix the type problem
+            # but the code should be correct, and is tested
+            if not issubclass(token_provider, TokenProvider):  # type: ignore
+                raise err
+        except (
+            TypeError
+        ):  # This is the error if token_provider is not a class for issubclass
+            raise err
+
+    def _credential_login(
+        self, username: str, password: Optional[str]
+    ) -> TokenProvider:
+        if password is None:
+            password = getpass("Enter your password:")
+
+        if not password:
+            raise ValueError("You cannot provide an empty password.")
```

Comment:
This is where I meant that it would be nice, in the first `if` statement, to check `if not password`. It would catch someone giving an empty string `""` as the value and let them give the password with the `getpass` functionality.

Reply:
hmm, I see your point of view.

However, a lot of the usage of the cloud-sdk is/will be indirect.
So if we first have this feature I think it should be done in a manner which makes it more likely that people can use it.

In much other Pasqal code we avoid `Optional` and default e.g. strings to `""`, so this checking just for `None` will never be reached.